nil pointer dereference with kuttl 0.4.0

I've just run into the following nil pointer crash and thought I'd report it:

```
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry | Creating namespace: kudo-test-poetic-troll
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry/0-create | starting test step 0-create
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry/0-create | running command: "helm init --client-only"
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry/0-create | running command: [helm init --client-only]
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry/0-create | $HELM_HOME has been configured at /tekton/home/.helm.
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry/0-create | Not installing Tiller due to 'client-only' flag having been set
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry/0-create | running command: "helm delete --purge docker-registry"
    kuttl/harness/registry: logger.go:42: 17:25:15 | registry/0-create | running command: [helm delete --purge docker-registry]
    kuttl/harness/registry: logger.go:42: 17:25:26 | registry/0-create | Error: release: "docker-registry" not found
...
    kuttl/harness/registry: logger.go:42: 17:26:52 | registry/0-create | Error: release docker-registry failed: Get https://10.96.0.1:443/api/v1/namespaces/kudo-test-poetic-troll: dial tcp 10.96.0.1:443: connect: connection refused
    kuttl/harness/registry: case.go:156: exit status 1
    kuttl/harness/registry: case.go:156: Get "https://127.0.0.1:32768/api?timeout=32s": EOF
    kuttl/harness/registry: case.go:159: failed in step 0-create
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x131e945]

goroutine 8260 [running]:
github.com/kudobuilder/kuttl/pkg/test/utils.(*RetryClient).List.func1(0x1936320, 0xc000046048, 0x0, 0x0)
	/Users/kensipe/projects/go/src/github.com/kudobuilder/kuttl/pkg/test/utils/kubernetes.go:228 +0x25
github.com/kudobuilder/kuttl/pkg/test/utils.Retry.func1(0xc0004670c0, 0x1936320, 0xc000046048, 0xc0006d5b60, 0xc0006d5ce0)
	/Users/kensipe/projects/go/src/github.com/kudobuilder/kuttl/pkg/test/utils/kubernetes.go:110 +0x3e
created by github.com/kudobuilder/kuttl/pkg/test/utils.Retry
	/Users/kensipe/projects/go/src/github.com/kudobuilder/kuttl/pkg/test/utils/kubernetes.go:109 +0xf7
Collecting kind cluster logs.
```

I believe this was with kuttl version 0.4.0.

--- a/pkg/test/utils/kubernetes.go
+++ b/pkg/test/utils/kubernetes.go
@@ -108,7 +108,18 @@ func Retry(ctx context.Context, fn func(context.Context) error, errValidationFun
 		// we can use select to wait for both the function return and the context deadline.
 
 		go func() {
-			if err := fn(ctx); err != nil {
+			defer func() {
+				if r := recover(); r != nil {
+					log.Println("retry func has panicked and will be ignored")
+					doneCh <- struct{}{}
+				}
+			}()
+
+			if fn == nil {
+				// fn is passed in usually
+				log.Println("retry func is nil and will be ignored")
+				doneCh <- struct{}{}
+			} else if err := fn(ctx); err != nil {
 				errCh <- err
 			} else {
 				doneCh <- struct{}{}

--- a/pkg/test/utils/kubernetes.go
+++ b/pkg/test/utils/kubernetes.go
@@ -117,8 +117,8 @@ func Retry(ctx context.Context, fn func(context.Context) error, errValidationFun
 			// the common case is when a shared refernence, like a client, is nil and is called in the function
 			defer func() {
 				if r := recover(); r != nil {
-					log.Println("retry func has panicked and will be ignored")
-					doneCh <- struct{}{}
+					//log.Println("retry func has panicked and will be ignored")
+					errCh <- errors.New("func passed to retry panicked.  expected if testsuite is shutting down")
 				}
 			}()
 

--- a/pkg/test/utils/kubernetes.go
+++ b/pkg/test/utils/kubernetes.go
@@ -114,7 +114,7 @@ func Retry(ctx context.Context, fn func(context.Context) error, errValidationFun
 
 		go func() {
 			// if the func we are calling panics, clean up and call it done
-			// the common case is when a shared refernence, like a client, is nil and is called in the function
+			// the common case is when a shared reference, like a client, is nil and is called in the function
 			defer func() {
 				if r := recover(); r != nil {
 					//log.Println("retry func has panicked and will be ignored")

--- a/pkg/test/utils/kubernetes.go
+++ b/pkg/test/utils/kubernetes.go
@@ -102,12 +102,19 @@ func Retry(ctx context.Context, fn func(context.Context) error, errValidationFun
 	errCh := make(chan error)
 	doneCh := make(chan struct{})
 
+	if fn == nil {
+		log.Println("retry func is nil and will be ignored")
+		return nil
+	}
+
 	// do { } while (err != nil): https://stackoverflow.com/a/32844744/10892393
 	for ok := true; ok; ok = lastErr != nil {
 		// run the function in a goroutine and close it once it is finished so that
 		// we can use select to wait for both the function return and the context deadline.
 
 		go func() {
+			// if the func we are calling panics, clean up and call it done
+			// the common case is when a shared refernence, like a client, is nil and is called in the function
 			defer func() {
 				if r := recover(); r != nil {
 					log.Println("retry func has panicked and will be ignored")
@@ -115,11 +122,7 @@ func Retry(ctx context.Context, fn func(context.Context) error, errValidationFun
 				}
 			}()
 
-			if fn == nil {
-				// fn is passed in usually
-				log.Println("retry func is nil and will be ignored")
-				doneCh <- struct{}{}
-			} else if err := fn(ctx); err != nil {
+			if err := fn(ctx); err != nil {
 				errCh <- err
 			} else {
 				doneCh <- struct{}{}

--- a/pkg/test/utils/kubernetes_test.go
+++ b/pkg/test/utils/kubernetes_test.go
@@ -131,7 +131,7 @@ func TestRetryWithNilFromFn(t *testing.T) {
 func TestRetryWithNilInFn(t *testing.T) {
 	client := RetryClient{}
 	var list runtime.Object
-	assert.Equal(t, nil, Retry(context.TODO(), func(ctx context.Context) error {
+	assert.Error(t, Retry(context.TODO(), func(ctx context.Context) error {
 		return client.Client.List(ctx, list)
 	}, IsJSONSyntaxError))
 }

--- a/pkg/test/utils/kubernetes_test.go
+++ b/pkg/test/utils/kubernetes_test.go
@@ -128,7 +128,7 @@ func TestRetryWithNilFromFn(t *testing.T) {
 	}, IsJSONSyntaxError))
 }
 
-func TestRetryWithNilWithFn(t *testing.T) {
+func TestRetryWithNilInFn(t *testing.T) {
 	client := RetryClient{}
 	var list runtime.Object
 	assert.Equal(t, nil, Retry(context.TODO(), func(ctx context.Context) error {

--- a/pkg/test/utils/kubernetes_test.go
+++ b/pkg/test/utils/kubernetes_test.go
@@ -118,6 +118,24 @@ func TestRetryWithUnexpectedError(t *testing.T) {
 	assert.Equal(t, 1, index)
 }
 
+func TestRetryWithNil(t *testing.T) {
+	assert.Equal(t, nil, Retry(context.TODO(), nil, IsJSONSyntaxError))
+}
+
+func TestRetryWithNilFromFn(t *testing.T) {
+	assert.Equal(t, nil, Retry(context.TODO(), func(ctx context.Context) error {
+		return nil
+	}, IsJSONSyntaxError))
+}
+
+func TestRetryWithNilWithFn(t *testing.T) {
+	client := RetryClient{}
+	var list runtime.Object
+	assert.Equal(t, nil, Retry(context.TODO(), func(ctx context.Context) error {
+		return client.Client.List(ctx, list)
+	}, IsJSONSyntaxError))
+}
+
 func TestRetryWithTimeout(t *testing.T) {
 	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
 	defer cancel()

